src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,366 @@
+				Flags: parsed.Flags,
+			}, true, nil
+		}
+	case syntax.OpEndText:
+		if rightmost {
+			return &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}, true, nil
+		}
+	case syntax.OpCapture:
+		// because golang regexp's don't allow backreferences, we don't care about maintaining capture
+		// group namings and can treate captures the same as we do conactenations.
+		fallthrough
+	case syntax.OpConcat:
+		changed := false
+		// strip left-most '^'
+		if l := len(parsed.Sub); leftmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost && l == 1)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+		return parsed, changed, nil
+	case syntax.OpAlternate:
+		changed := false
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[idx] = newRe
+				changed = true
+			}
+		}
+		return parsed, changed, nil
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				return parsed, true, nil
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}, true, nil
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
+	case syntax.OpRepeat:
+		if len(parsed.Sub) > 0 && parsed.Min > 0 && parsed.Max > 0 {
```

Comment:
Still not sure about this case, don't you just want `parsed.Min > 0`?  When `parsed.Max == -1` which means unlimited range (i.e. `{2,}`) you still want to do this no?

Reply:
This is the case I'm thinking of:
https://play.golang.org/p/O0uAkMV-pEr